MigrationAwareServiceEventTest.partitionIsMigratingFlag_shouldBeSet_until_serviceCommitRollback_isCompleted

```
java.lang.AssertionError: Unexpected responses: [true]
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.internal.partition.MigrationAwareServiceEventTest.partitionIsMigratingFlag_shouldBeSet_until_serviceCommitRollback_isCompleted(MigrationAwareServiceEventTest.java:109)
```

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6234/testReport/junit/com.hazelcast.internal.partition/MigrationAwareServiceEventTest/partitionIsMigratingFlag_shouldBeSet_until_serviceCommitRollback_isCompleted/

```
java.lang.AssertionError: 
Expected: an empty collection
     but: <[Response: true, Event: PartitionMigrationEvent{migrationEndpoint=DESTINATION, partitionId=46, currentReplicaIndex=1, newReplicaIndex=0}, Type: COMMIT]>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at com.hazelcast.internal.partition.MigrationAwareServiceEventTest.partitionIsMigratingFlag_shouldBeSet_until_serviceCommitRollback_isCompleted(MigrationAwareServiceEventTest.java:109)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK8/com.hazelcast$hazelcast/423/testReport/junit/com.hazelcast.internal.partition/MigrationAwareServiceEventTest/partitionIsMigratingFlag_shouldBeSet_until_serviceCommitRollback_isCompleted/

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateManager.java
@@ -266,10 +266,16 @@ public class PartitionStateManager {
     }
 
     public void setMigratingFlag(int partitionId) {
+        if (logger.isFinestEnabled()) {
+            logger.finest("Setting partition-migrating flag. partitionId=" + partitionId);
+        }
         partitions[partitionId].setMigrating(true);
     }
 
     public void clearMigratingFlag(int partitionId) {
+        if (logger.isFinestEnabled()) {
+            logger.finest("Clearing partition-migrating flag. partitionId=" + partitionId);
+        }
         partitions[partitionId].setMigrating(false);
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BeforePromotionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BeforePromotionOperation.java
@@ -16,6 +16,8 @@
 
 package com.hazelcast.internal.partition.operation;
 
+import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
+import com.hazelcast.internal.partition.impl.PartitionStateManager;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.spi.MigrationAwareService;
 import com.hazelcast.spi.PartitionMigrationEvent;
@@ -26,6 +28,8 @@ import static com.hazelcast.core.MigrationEvent.MigrationStatus.STARTED;
 // before applying promotion result to the partition table.
 final class BeforePromotionOperation extends AbstractPromotionOperation {
 
+    private Runnable beforePromotionsCallback;
+
     /**
      * This constructor should not be used to obtain an instance of this class; it exists to fulfill IdentifiedDataSerializable
      * coding conventions.
@@ -34,8 +38,9 @@ final class BeforePromotionOperation extends AbstractPromotionOperation {
         super(-1);
     }
 
-    BeforePromotionOperation(int currentReplicaIndex) {
+    BeforePromotionOperation(int currentReplicaIndex, Runnable beforePromotionsCallback) {
         super(currentReplicaIndex);
+        this.beforePromotionsCallback = beforePromotionsCallback;
     }
 
     @Override
@@ -59,8 +64,15 @@ final class BeforePromotionOperation extends AbstractPromotionOperation {
             try {
                 service.beforeMigration(event);
             } catch (Throwable e) {
-                logger.warning("While promoting partitionId=" + getPartitionId(), e);
+                logger.warning("While promoting " + getPartitionMigrationEvent(), e);
             }
+        }
+    }
+
+    @Override
+    public void afterRun() throws Exception {
+        if (beforePromotionsCallback != null) {
+            beforePromotionsCallback.run();
         }
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BeforePromotionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BeforePromotionOperation.java
@@ -41,6 +41,10 @@ final class BeforePromotionOperation extends AbstractPromotionOperation {
     @Override
     public void beforeRun() throws Exception {
         sendMigrationEvent(STARTED);
+
+        InternalPartitionServiceImpl service = getService();
+        PartitionStateManager partitionStateManager = service.getPartitionStateManager();
+        partitionStateManager.setMigratingFlag(getPartitionId());
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizePromotionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizePromotionOperation.java
@@ -74,9 +74,11 @@ final class FinalizePromotionOperation extends AbstractPromotionOperation {
 
     @Override
     public void afterRun() throws Exception {
-        clearPartitionMigratingFlag();
-        MigrationEvent.MigrationStatus status = success ? COMPLETED : FAILED;
-        sendMigrationEvent(status);
+        InternalPartitionServiceImpl service = getService();
+        PartitionStateManager partitionStateManager = service.getPartitionStateManager();
+        partitionStateManager.clearMigratingFlag(getPartitionId());
+
+        sendMigrationEvent(success ? COMPLETED : FAILED);
     }
 
     private void shiftUpReplicaVersions() {
@@ -130,11 +132,5 @@ final class FinalizePromotionOperation extends AbstractPromotionOperation {
                 logger.warning("While promoting partitionId=" + getPartitionId(), e);
             }
         }
-    }
-
-    private void clearPartitionMigratingFlag() {
-        InternalPartitionServiceImpl service = getService();
-        PartitionStateManager partitionStateManager = service.getPartitionStateManager();
-        partitionStateManager.clearMigratingFlag(getPartitionId());
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizePromotionOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/FinalizePromotionOperation.java
@@ -16,7 +16,6 @@
 
 package com.hazelcast.internal.partition.operation;
 
-import com.hazelcast.core.MigrationEvent;
 import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
 import com.hazelcast.internal.partition.impl.PartitionEventManager;
 import com.hazelcast.internal.partition.impl.PartitionStateManager;
@@ -118,7 +117,7 @@ final class FinalizePromotionOperation extends AbstractPromotionOperation {
             try {
                 service.commitMigration(event);
             } catch (Throwable e) {
-                logger.warning("While promoting partitionId=" + getPartitionId(), e);
+                logger.warning("While promoting " + getPartitionMigrationEvent(), e);
             }
         }
     }
@@ -129,7 +128,7 @@ final class FinalizePromotionOperation extends AbstractPromotionOperation {
             try {
                 service.rollbackMigration(event);
             } catch (Throwable e) {
-                logger.warning("While promoting partitionId=" + getPartitionId(), e);
+                logger.warning("While promoting " + getPartitionMigrationEvent(), e);
             }
         }
     }

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java
@@ -24,7 +24,6 @@ import com.hazelcast.internal.partition.MigrationInfo;
 import com.hazelcast.internal.partition.PartitionRuntimeState;
 import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
 import com.hazelcast.internal.partition.impl.PartitionDataSerializerHook;
-import com.hazelcast.internal.partition.impl.PartitionStateManager;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
@@ -33,7 +32,6 @@ import com.hazelcast.spi.NodeEngine;
 import com.hazelcast.spi.OperationService;
 import com.hazelcast.spi.exception.TargetNotMemberException;
 import com.hazelcast.spi.impl.NodeEngineImpl;
-import com.hazelcast.spi.impl.PartitionSpecificRunnable;
 import com.hazelcast.spi.impl.operationservice.InternalOperationService;
 import com.hazelcast.util.Preconditions;
 
@@ -106,25 +104,26 @@ public class PromotionCommitOperation extends AbstractPartitionOperation impleme
         NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
         InternalOperationService operationService = nodeEngine.getOperationService();
         InternalPartitionServiceImpl partitionService = getService();
-        PartitionStateManager partitionStateManager = partitionService.getPartitionStateManager();
-        AtomicInteger tasks = new AtomicInteger(promotions.size());
 
         ILogger logger = getLogger();
-        if (logger.isFinestEnabled()) {
-            logger.finest("Submitting before promotion tasks for " + promotions);
-        } else if (logger.isFineEnabled()) {
-            logger.fine("Submitting before promotion tasks for " + promotions.size() + " promotions.");
+        if (logger.isFineEnabled()) {
+            logger.fine("Submitting BeforePromotionOperations for " + promotions.size() + " promotions.");
         }
+
+        Runnable beforePromotionsCallback = new BeforePromotionOperationCallback(this, new AtomicInteger(promotions.size()));
+
         for (MigrationInfo promotion : promotions) {
-            operationService.execute(new BeforePromotionTask(this, promotion, nodeEngine, tasks));
+            if (logger.isFinestEnabled()) {
+                logger.finest("Submitting BeforePromotionOperation for promotion: " + promotion);
+            }
+            int currentReplicaIndex = promotion.getDestinationCurrentReplicaIndex();
+            BeforePromotionOperation op = new BeforePromotionOperation(currentReplicaIndex, beforePromotionsCallback);
+            op.setPartitionId(promotion.getPartitionId()).setNodeEngine(nodeEngine).setService(partitionService);
+            operationService.execute(op);
         }
     }
 
     private void finalizePromotion() {
-        ILogger logger = getLogger();
-        if (logger.isFineEnabled()) {
-            logger.fine("Running promotion finalization for " + promotions.size() + " promotions.");
-        }
         NodeEngine nodeEngine = getNodeEngine();
         InternalPartitionServiceImpl partitionService = getService();
         OperationService operationService = nodeEngine.getOperationService();
@@ -132,12 +131,14 @@ public class PromotionCommitOperation extends AbstractPartitionOperation impleme
         partitionState.setEndpoint(getCallerAddress());
         success = partitionService.processPartitionRuntimeState(partitionState);
 
-        if (logger.isFinestEnabled()) {
-            logger.finest("Submitting finalize promotion operations for " + promotions);
-        } else if (logger.isFineEnabled()) {
-            logger.fine("Submitting finalize promotion operations for " + promotions.size() + " promotions.");
+        ILogger logger = getLogger();
+        if (logger.isFineEnabled()) {
+            logger.fine("Submitting FinalizePromotionOperations for " + promotions.size() + " promotions. Result: " + success);
         }
         for (MigrationInfo promotion : promotions) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Submitting FinalizePromotionOperation for promotion: " + promotion + ". Result: " + success);
+            }
             int currentReplicaIndex = promotion.getDestinationCurrentReplicaIndex();
             FinalizePromotionOperation op = new FinalizePromotionOperation(currentReplicaIndex, success);
             op.setPartitionId(promotion.getPartitionId()).setNodeEngine(nodeEngine).setService(partitionService);
@@ -150,54 +151,34 @@ public class PromotionCommitOperation extends AbstractPartitionOperation impleme
         return PartitionDataSerializerHook.PROMOTION_COMMIT;
     }
 
-    private static class BeforePromotionTask implements PartitionSpecificRunnable {
+    private static class BeforePromotionOperationCallback implements Runnable {
         private final PromotionCommitOperation promotionCommitOperation;
-        private final MigrationInfo promotion;
-        private final NodeEngineImpl nodeEngine;
         private final AtomicInteger tasks;
 
-        BeforePromotionTask(PromotionCommitOperation promotionCommitOperation, MigrationInfo promotion,
-                NodeEngineImpl nodeEngine, AtomicInteger tasks) {
+        BeforePromotionOperationCallback(PromotionCommitOperation promotionCommitOperation, AtomicInteger tasks) {
             this.promotionCommitOperation = promotionCommitOperation;
-            this.promotion = promotion;
-            this.nodeEngine = nodeEngine;
             this.tasks = tasks;
         }
 
         @Override
         public void run() {
-            try {
-                int currentReplicaIndex = promotion.getDestinationCurrentReplicaIndex();
-                BeforePromotionOperation op = new BeforePromotionOperation(currentReplicaIndex);
-                op.setPartitionId(promotion.getPartitionId()).setNodeEngine(nodeEngine)
-                        .setService(nodeEngine.getPartitionService());
-
-                InternalOperationService operationService = nodeEngine.getOperationService();
-                operationService.run(op);
-            } finally {
-                completeTask();
-            }
-        }
-
-        private void completeTask() {
             final int remainingTasks = tasks.decrementAndGet();
 
-            ILogger logger = nodeEngine.getLogger(getClass());
+            ILogger logger = promotionCommitOperation.getLogger();
             if (logger.isFinestEnabled()) {
                 logger.finest("Remaining before promotion tasks: " + remainingTasks);
             }
 
             if (remainingTasks == 0) {
-                logger.fine("All before promotion tasks are completed, re-submitting PromotionCommitOperation.");
-                promotionCommitOperation.beforeStateCompleted = true;
-                nodeEngine.getOperationService().execute(promotionCommitOperation);
+                logger.fine("All before promotion tasks are completed.");
+                promotionCommitOperation.onBeforePromotionsComplete();
             }
         }
+    }
 
-        @Override
-        public int getPartitionId() {
-            return promotion.getPartitionId();
-        }
+    private void onBeforePromotionsComplete() {
+        beforeStateCompleted = true;
+        getNodeEngine().getOperationService().execute(this);
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java
@@ -116,7 +116,6 @@ public class PromotionCommitOperation extends AbstractPartitionOperation impleme
             logger.fine("Submitting before promotion tasks for " + promotions.size() + " promotions.");
         }
         for (MigrationInfo promotion : promotions) {
-            partitionStateManager.setMigratingFlag(promotion.getPartitionId());
             operationService.execute(new BeforePromotionTask(this, promotion, nodeEngine, tasks));
         }
     }

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/MigrationAwareServiceEventTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/MigrationAwareServiceEventTest.java
@@ -19,6 +19,7 @@ package com.hazelcast.internal.partition;
 import com.hazelcast.config.Config;
 import com.hazelcast.config.ServiceConfig;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl;
 import com.hazelcast.spi.ManagedService;
 import com.hazelcast.spi.MigrationAwareService;
 import com.hazelcast.spi.NodeEngine;
@@ -123,9 +124,11 @@ public class MigrationAwareServiceEventTest extends HazelcastTestSupport {
         @Override
         public void sendResponse(Operation operation, Object response) {
             assert operation instanceof DummyPartitionAwareOperation : "Invalid operation: " + operation;
-            if (!(response instanceof RetryableHazelcastException)) {
+            NodeEngine nodeEngine = operation.getNodeEngine();
+            if (!(response instanceof RetryableHazelcastException) && nodeEngine.isRunning()) {
                 DummyPartitionAwareOperation op = (DummyPartitionAwareOperation) operation;
-                failures.add("Response: " + response + ", Event: " + op.event + ", Type: " + op.type);
+                failures.add("Unexpected response: " + response + ". Node: " + nodeEngine.getThisAddress()
+                        + ", Event: " + op.event + ", Type: " + op.type);
             }
         }
     }
@@ -158,24 +161,45 @@ public class MigrationAwareServiceEventTest extends HazelcastTestSupport {
 
         @Override
         public void commitMigration(PartitionMigrationEvent event) {
-            executePartitionOperation(event, TYPE_COMMIT);
-        }
-
-        private void executePartitionOperation(PartitionMigrationEvent event, String type) {
-            if (event.getNewReplicaIndex() != 0 && event.getCurrentReplicaIndex() != 0) {
-                return;
-            }
-
-            DummyPartitionAwareOperation op = new DummyPartitionAwareOperation(event, type);
-            int replicaIndex = type.equals(TYPE_COMMIT) ? event.getNewReplicaIndex() : event.getCurrentReplicaIndex();
-            op.setPartitionId(event.getPartitionId()).setReplicaIndex(replicaIndex);
-            op.setOperationResponseHandler(responseHandler);
-            nodeEngine.getOperationService().run(op);
+            checkPartition(event, TYPE_COMMIT);
         }
 
         @Override
         public void rollbackMigration(PartitionMigrationEvent event) {
-            executePartitionOperation(event, TYPE_ROLLBACK);
+            checkPartition(event, TYPE_ROLLBACK);
+        }
+
+        private void checkPartition(PartitionMigrationEvent event, String type) {
+            if (event.getNewReplicaIndex() != 0 && event.getCurrentReplicaIndex() != 0) {
+                return;
+            }
+
+            checkPartitionMigrating(event, type);
+
+            if (event.getCurrentReplicaIndex() != -1) {
+                runPartitionOperation(event, type, event.getCurrentReplicaIndex());
+            }
+            if (event.getNewReplicaIndex() != -1) {
+                runPartitionOperation(event, type, event.getNewReplicaIndex());
+            }
+        }
+
+        private void runPartitionOperation(PartitionMigrationEvent event, String type, int replicaIndex) {
+            DummyPartitionAwareOperation op = new DummyPartitionAwareOperation(event, type);
+            op.setNodeEngine(nodeEngine).setPartitionId(event.getPartitionId()).setReplicaIndex(replicaIndex);
+            op.setOperationResponseHandler(responseHandler);
+            nodeEngine.getOperationService().run(op);
+        }
+
+        private void checkPartitionMigrating(PartitionMigrationEvent event, String type) {
+            InternalPartitionServiceImpl partitionService =
+                    (InternalPartitionServiceImpl) nodeEngine.getPartitionService();
+
+            InternalPartition partition = partitionService.getPartition(event.getPartitionId());
+            if (!partition.isMigrating() && nodeEngine.isRunning()) {
+                responseHandler.failures.add("Migrating flag is not set. Node: " + nodeEngine.getThisAddress()
+                        + ", Event: " + event + ", Type: " + type);
+            }
         }
 
         @Override

